### PR TITLE
Connect API to proxy with REST API

### LINKED ISSUE #3413: REST API: Add to Catalog API ability to connect an API to proxy
## Description
- "As an API owner I want to connect my API to proxy using  REST Catalog API"
- "As an API owner I want to disconnect my API from proxy using REST Catalog API"
- "As an API owner I want to see endpoint and base addresses of my API, when it is connected to a proxy"

### New endpoints
~~ #### GET /proxies
- response contains list of available proxies (name, id, type) ~~ see Update: Proxy API

#### POST /apis/:id/proxy
Parameters (in URL)
- :id
  - Expected value: id of API in question

Parameter (in body)
-  "proxyId"
   -  expected value:  id of an existing proxy
-  "frontendPrefix", which is a string to be added to all proxy requests (identifies the API by proxy)
   - Expected value: a unique string
-  "backendPrefix", which is an identification of API on server
   - Expected value: a string 
- "apiPort", port number on API server
   - Expected value: a number within port number space
- "disableApiKey", whether API key is required or not
   - Expected value: "true" | "false"  

Functionality
  - if proxy with given proxy-id is found, the API is connected to that proxy
    - the newly created proxyBackend object is returned in response.
  - if proxy is not found, error response with code 400 is returned
  - if the API was already connected to a proxy, error response  with code 400 is returned
 
#### DELETE /apis/:id/proxy/:id
- if proxy backend with given proxy-id and id of API in question is found, the API is disconnected from proxy
- if proxy backend is not found, error response with code 400 is returned

#### GET /apis (and similar responses)
- when Admin is requesting API data, in addition to parameter url, return also
  - API backend URL (the address, which was given, when API was created)

#### GET /apis /:id 
- when Admin is requesting API data,  return also proxy backend data, if API is connected to a proxy

### PR DESCRIPTION
**Pull request by matleppa**
Closes #3413 

## Changes
- In case API is connected to a proxy, manager sees also backend URL, backend prefix, proxy name and proxy type in API data 
- ability to connect and disconnect an API to/from apiUmbrella proxy
- ability to register and unregister an API as a publisher or a subscriber to/from an MQTT broker

## Developer checklist
This checklist is to be completed by the PR developer:

- [ ] Alternative solutions were compared/discussed before writing code
    - [ ] trade-offs with this solution are considered acceptable
- [ ] Code in this PR adheres to the [project styleguide](CONTRIBUTING.md)
- [ ] This pull request does not decrease project test coverage
- [ ] If the code changes existing database collection(s), migration has been written
- [ ] If UI texts are added or changed, all texts are internationalized

## Reviewer checklist
Reviewed by: @username1

This list is to be completed by the pull request reviewer:
- [ ] Code works as described/expected
- [ ] Code seems to be error free
    - [ ] no browser console errors visible
    - [ ] no server console errors visible
    - [ ] passes CI build
- [ ] Code is written in a way that promotes maintainability
    - [ ] easy to understand
    - [ ] well organized
    - [ ] follows project coding standards and conventions
    - [ ] well documented

### CODE REVIEW

**Comment by marla-singer** [comment]
@matleppa I suggest to create a new Swagger documentation to manage Proxy & Proxy Backend with name `proxy.json`/ What do you think?

**Comment by matleppa** [comment]
@marla-singer Good point! 
After some consideration I think that it will indeed be feasible to create a new internal API to manage the proxies related operations. That will be used for similar task as the Proxies in UI: to create, modify and remove proxies.

However, I think that Proxy Backend related functionality belongs to Catalog API, where also other API related operations are performed, pretty much same way as it is separated in UI currently.
Do you agree on that?

**Comment by marla-singer** [comment]
> However, I think that Proxy Backend related functionality belongs to Catalog API, where also other API related operations are performed, pretty much same way as it is separated in UI currently.
Do you agree on that?

Totally agree. 

And I suppose  it would be a new endpoint as /api/{id}/proxy or like that for management Proxy backend

**Comment by matleppa** [comment]
> And  I suppose it would be a new endpoint as /api/{id}/proxy or like that for management Proxy backend

Exactly!

File: apinf_packages/proxies/server/api.js [hunk 1]
```diff
@@ -0,0 +1,853 @@
...
+};
+
+// Request /rest/v1/proxies for Proxies collection
+ProxyV1.addCollection(Proxies, {
+  routeOptions: {
+    authRequired: false,
+  },
+  endpoints: {
+    // Response contains a list of all public entities within the collection
+    getAll: {
```
Comment (marla-singer) [inline apinf_packages/proxies/server/api.js]: About endpoints that return all Proxies. I suggest to add query parameter as "type" with values "emq" and "api-umbrella". It'd provide all proxies with specific type. If type is not provided that returns all

@matleppa What do you think?

File: apinf_packages/proxies/server/api.js [hunk 2]
```diff
@@ -0,0 +1,853 @@
...
+          body: {
+            status: 'success',
+            data: proxyList,
+          },
+        };
+      },
+    },
+    // Response contains the entity with the given :id
+    get: {
+      authRequired: false,
```
Comment (marla-singer) [inline apinf_packages/proxies/server/api.js]: Should be `authRequired: false` & `roleRequired: ['admin'],`

File: apinf_packages/proxies/server/api.js [hunk 3]
```diff
@@ -0,0 +1,853 @@
...
+                  type: 'string',
+                  example: 'success',
+                },
+                data: {
+                  $ref: '#/definitions/proxyResponse',
+                },
+              },
+            },
+          },
+          204: {
```
Comment (marla-singer) [inline apinf_packages/proxies/server/api.js]: Why do we need 204 status for GET request?

File: apinf_packages/proxies/server/api.js [hunk 4]
```diff
@@ -0,0 +1,853 @@
...
+      swagger: {
+        tags: [
+          ProxyV1.swagger.tags.proxy,
+        ],
+        summary: 'Fetch Proxy with specified ID.',
+        description: descriptionProxies.getProxy,
+        parameters: [
+          ProxyV1.swagger.params.proxyId,
+        ],
+        responses: {
```
Comment (marla-singer) [inline apinf_packages/proxies/server/api.js]: 400, 403 statuses are missed in the specification

File: apinf_packages/proxies/server/api.js [hunk 5]
```diff
@@ -0,0 +1,853 @@
...
+      swagger: {
+        tags: [
+          ProxyV1.swagger.tags.proxy,
+        ],
+        summary: 'Add a new Proxy.',
+        description: descriptionProxies.postProxy,
+        parameters: [
+          ProxyV1.swagger.params.proxyRequest,
+        ],
+        responses: {
```
Comment (marla-singer) [inline apinf_packages/proxies/server/api.js]: 403 status is missed in the specification

File: apinf_packages/proxies/server/api.js [hunk 6]
```diff
@@ -0,0 +1,853 @@
...
+        security: [
+          {
+            userSecurityToken: [],
+            userId: [],
+          },
+        ],
+      },
+      action () {
+        const userId = this.userId;
+        const bodyParams = this.bodyParams;
```
Comment (marla-singer) [inline apinf_packages/proxies/server/api.js]: The block with checking user permission is missed

File: apinf_packages/proxies/server/api.js [hunk 7]
```diff
@@ -0,0 +1,853 @@
...
+        tags: [
+          ProxyV1.swagger.tags.proxy,
+        ],
+        summary: 'Update Proxy.',
+        description: descriptionProxies.putProxy,
+        parameters: [
+          ProxyV1.swagger.params.proxyId,
+          ProxyV1.swagger.params.proxyRequest,
+        ],
+        responses: {
```
Comment (marla-singer) [inline apinf_packages/proxies/server/api.js]: I guess the 500 status is missed in the specification

File: apinf_packages/proxies/server/api.js [hunk 8]
```diff
@@ -0,0 +1,853 @@
...
+      swagger: {
+        tags: [
+          ProxyV1.swagger.tags.proxy,
+        ],
+        summary: 'Delete API.',
+        description: descriptionProxies.deleteProxy,
+        parameters: [
+          ProxyV1.swagger.params.proxyId,
+        ],
+        responses: {
```
Comment (marla-singer) [inline apinf_packages/proxies/server/api.js]: 500 status is missed in the specification

File: apinf_packages/rest_apis/server/proxy.js [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+    },
+    url: {
+      name: 'url',
+      in: 'query',
+      description: 'Documentation link to be removed.',
+      required: false,
+      type: 'string',
+    },
+
+    // Proxy related parameters
```
Comment (marla-singer) [inline apinf_packages/rest_apis/server/proxy.js]: needless an empty line

File: apinf_packages/rest_apis/server/proxy.js [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+    proxyRequest: {
+      name: 'proxy',
+      in: 'body',
+      description: 'Data for adding or editing a Proxy',
+      schema: {
+        $ref: '#/definitions/proxyRequest',
+      },
+    },
+
+  },
```
Comment (marla-singer) [inline apinf_packages/rest_apis/server/proxy.js]: Needless an empty line

File: apinf_packages/rest_apis/server/proxy.js [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+          type: 'string',
+          enum: ['true', 'false'],
+          example: 'true | false',
+        },
+        emqHttpApi: {
+          type: 'string',
+          format: 'url',
+          example: 'https://emq.httpapi.com',
+        },
+
```
Comment (marla-singer) [inline apinf_packages/rest_apis/server/proxy.js]: Needless an empty line

File: apinf_packages/rest_apis/server/proxy.js [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+            userId: {
+              type: 'string',
+              example: 'user-id-value',
+            },
+          },
+        },
+      },
+    },
+
+    // proxy related definitions
```
Comment (marla-singer) [inline apinf_packages/rest_apis/server/proxy.js]: Needless an empty line

File: apinf_packages/rest_apis/server/proxy.js [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+            },
+          },
+        },
+      },
+    },
+  },
+};
+
+// Generate Swagger to route /rest/v1/api_catalog.json
+ProxyV1.addSwagger('proxy.json');
```
Comment (marla-singer) [inline apinf_packages/rest_apis/server/proxy.js]: Update comment
`// Generate Swagger to route /rest/v1/proxy.json`

File: apinf_packages/proxies/server/api.js (hunk 2 above)
Comment (marla-singer) [reply apinf_packages/proxies/server/api.js]: Security block is missed for this endpoint
```
security: [
+          {
+            userSecurityToken: [],
+            userId: [],
+          },
+        ],
```

File: apinf_packages/proxies/server/api.js (hunk 1 above)
Comment (matleppa) [reply apinf_packages/proxies/server/api.js]: @marla-singer Good point. I agree.

**Comment by matleppa** [comment]
Replaced with PR 3547.

File: .meteor/packages [hunk 14]
```diff
@@ -96,3 +93,5 @@ zimme:active-route
 aldeed:autoform-select2
 natestrauser:select2
 meteorhacks:aggregate
+apinf:api-umbrella
```
[the remark this answers is not in the document]
Comment (matleppa) [reply .meteor/packages]: Not related to this PR.

File: .meteor/packages [hunk 15]
```diff
@@ -86,7 +84,6 @@ tap:i18n-ui
 tmeasday:publish-counts
 todda00:friendly-slugs
 tracker@1.1.3
-twbs:bootstrap
```
[the remark this answers is not in the document]
Comment (matleppa) [reply .meteor/packages]: Not related to this PR.

**Pull request closed without merging**